awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx

```diff
@@ -0,0 +1,167 @@
+import React from 'react';
+import { useField, useFormikContext } from 'formik';
+import { shape, string } from 'prop-types';
+import { withI18n } from '@lingui/react';
+import { t } from '@lingui/macro';
+import {
+  FormGroup,
+  InputGroup,
+  TextArea,
+  TextInput,
+} from '@patternfly/react-core';
+import { FieldTooltip, PasswordInput } from '../../../../components/FormField';
+import AnsibleSelect from '../../../../components/AnsibleSelect';
+import { CredentialType } from '../../../../types';
+import { required } from '../../../../util/validators';
+import { CredentialPluginField } from './CredentialPlugins';
+import BecomeMethodField from './BecomeMethodField';
+
+function CredentialInput({ fieldOptions, credentialKind, ...rest }) {
+  const [subFormField, meta] = useField(`inputs.${fieldOptions.id}`);
+  const isValid = !(meta.touched && meta.error);
+  if (fieldOptions.multiline) {
+    return (
+      <TextArea
+        {...subFormField}
+        id={`credential-${fieldOptions.id}`}
+        rows={6}
+        resizeOrientation="vertical"
+        onChange={(value, event) => {
+          subFormField.onChange(event);
+        }}
+        isValid={isValid}
+      />
+    );
+  }
+  if (fieldOptions.secret) {
+    const passwordInput = () => (
+      <PasswordInput
+        {...subFormField}
+        id={`credential-${fieldOptions.id}`}
+        isValid={isValid}
+        {...rest}
+      />
+    );
+    return credentialKind === 'external' ? (
+      <InputGroup>{passwordInput()}</InputGroup>
+    ) : (
+      passwordInput()
+    );
+  }
+  return (
+    <TextInput
+      {...subFormField}
+      id={`credential-${fieldOptions.id}`}
+      onChange={(value, event) => {
+        subFormField.onChange(event);
+      }}
+      isValid={isValid}
+    />
+  );
+}
+
+CredentialInput.propTypes = {
+  fieldOptions: shape({
+    id: string.isRequired,
+    label: string.isRequired,
+  }).isRequired,
+  credentialKind: string,
+};
+
+CredentialInput.defaultProps = {
+  credentialKind: '',
+};
+
+function CredentialField({ credentialType, fieldOptions, i18n }) {
+  const { values: formikValues } = useFormikContext();
+  const requiredFields = credentialType?.inputs?.required || [];
+  const isRequired = requiredFields.includes(fieldOptions.id);
+  const [subFormField, meta, helpers] = useField({
+    name: `inputs.${fieldOptions.id}`,
+    validate:
+      isRequired && !formikValues?.passwordPrompts[fieldOptions.id]
```

Comment:
can we unroll this from the ternaries and && into a function?

Reply:
+1

Reply:
Sure, np

Reply:
Talked with @jlmitch5 about this one and I think we came up with more readable solution.